src/ParallelAlgorithms/NonlinearSolver/NewtonRaphson/ResidualMonitorActions.hpp

```diff
@@ -0,0 +1,211 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <algorithm>
+#include <cstddef>
+#include <variant>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/Prefixes.hpp"
+#include "Informer/Tags.hpp"
+#include "Informer/Verbosity.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "Parallel/Info.hpp"
+#include "Parallel/Invoke.hpp"
+#include "Parallel/Printf.hpp"
+#include "ParallelAlgorithms/NonlinearSolver/NewtonRaphson/LineSearch.hpp"
+#include "ParallelAlgorithms/NonlinearSolver/NewtonRaphson/Tags/InboxTags.hpp"
+#include "ParallelAlgorithms/NonlinearSolver/Observe.hpp"
+#include "ParallelAlgorithms/NonlinearSolver/Tags.hpp"
+#include "Utilities/EqualWithinRoundoff.hpp"
+#include "Utilities/Functional.hpp"
+#include "Utilities/GetOutput.hpp"
+#include "Utilities/Requires.hpp"
+
+/// \cond
+namespace tuples {
+template <typename...>
+class TaggedTuple;
+}  // namespace tuples
+/// \endcond
+
+namespace NonlinearSolver::newton_raphson::detail {
+
+template <typename FieldsTag, typename OptionsGroup, typename BroadcastTarget>
+struct CheckResidualMagnitude {
+  using fields_tag = FieldsTag;
+  using residual_tag =
+      db::add_tag_prefix<NonlinearSolver::Tags::Residual, fields_tag>;
+  using residual_magnitude_square_tag =
+      LinearSolver::Tags::MagnitudeSquare<residual_tag>;
+  using initial_residual_magnitude_tag =
+      ::Tags::Initial<LinearSolver::Tags::Magnitude<residual_tag>>;
+  using prev_residual_magnitude_square_tag =
+      NonlinearSolver::Tags::Globalization<residual_magnitude_square_tag>;
+
+  template <typename ParallelComponent, typename DataBox,
+            typename Metavariables, typename ArrayIndex,
+            Requires<db::tag_is_retrievable_v<residual_magnitude_square_tag,
```

Comment:
[optional] switch Requires to `if constexpr` to provide users with a better error message than "Simple action.... couldn't be called. Good luck, my friend."

Reply:
I added a SFINAE overload to emit an error message, since `if constexpr` would indent the entire action by one level